### ISSUE TITLE
Are Mobility.default_fallbacks reflecting I18n.fallbacks correctly?

### ISSUE THREAD

**Issue by ingemar** [shioyama/mobility#161]
I'm not suer that `Mobility.default_fallbacks` are reflecting `I18n.fallbacks correctly`, or at least it doesn't behave as I expect it to.

## Context
```ruby
# config/application.rb
config.i18n.default_locale = :sv
config.i18n.fallbacks = {
  en: %i[sv],
  de: %i[en sv],
}

# pry/irb
I18n.fallbacks[:sv]
# => [:sv]
I18n.fallbacks[:en]
# => [:en, :sv]
I18n.fallbacks[:de]
# => [:de, :en, :sv]
Mobility.default_fallbacks[:sv]
# => [:sv]
Mobility.default_fallbacks[:en]
# => [:en, :sv]
Mobility.default_fallbacks[:de]
# => [:de, :sv]
```

## Expected Behavior
I would expect that `Mobility.default_fallbacks` for `:de` would result in all three locales; i.e. `:de`, `:en`, and `:sv` – just like `I18n.fallbacks`.

## Actual Behavior
Second fallback `:en` omitted.

**Comment by shioyama** [comment]
Actually, `Mobility.default_fallbacks` defaults to `I18n::Fallbacks.new({})`. Maybe that should be changed.

**Comment by shioyama** [comment]
The method `default_fallbacks` has been renamed `new_fallbacks` in 0.4, the code is [here](https://github.com/shioyama/mobility/blob/429308d0943e603621a6dbbcf572178c8b2e414c/lib/mobility/configuration.rb#L55-L57). If no argument is passed, it defaults to `{}`, which is why you get `I18n::Fallbacks.new({})`. Perhaps with no parameters it should just return `I18n.fallbacks` instead.

**Comment by shioyama** [comment]
So I've looked a bit more closely at the code for `I18n::Locale::Fallbacks`, and honestly I'm really hesitant to do anything here that would couple Mobility more closely with I18n. Take a loook at [the code](https://github.com/svenfuchs/i18n/blob/7c6ccf471ecf2c0a406d3a5317375f28d2da3613/lib/i18n/locale/fallbacks.rb) for yourself, it's really hard to understand.

I also looked at Globalize, and what it does is to kind of "mix in" any fallbacks set through `Globalize.fallbacks =` with the defaults in `I18n.fallbacks`. You can see that code [here](https://github.com/globalize/globalize/blob/342a457548ce358f6fe5f7f5fdfeaae61560ccd2/lib/globalize.rb#L48-L50).

Mobility currently entirely ignores `I18n.fallbacks` altogether, such that each model gets its own instance of `I18n::Locale::Fallbacks`, with optional model-specific or global fallbacks assigned to it.

We could change that by mixing in the I18n defaults, so that it acts like Globalize. You can test this like this (in 0.4.0):

```ruby
I18n.config.fallbacks_generator = Proc.new do |fallbacks|
  I18n.fallbacks.dup.tap { |default_fallbacks| default_fallbacks.map(fallbacks) }
end
```

`map` here just adds the fallbacks passed to `fallbacks: ...` in `translates` on the model to a dup'ed copy of `I18n.fallbacks`. By doing it this way, if you set `fallbacks: ...` in a model, it would "merge" these fallbacks into the ones defined on `I18n.fallbacks`, which I suppose makes sense.

@pwim Any thoughts on this one? I never really used fallbacks in Globalize and don't have a good sense for what people use them for.

**Comment by shioyama** [comment]
Having said the above, the fact that `config.i18n.fallbacks` in Rails currently does not have any effect in Mobility is probably enough to justify the change in the default fallbacks generator above. Clearly people will expect that i18n fallbacks set in application config apply to Mobility.

**Comment by ingemar** [comment]
Hmm... Just looked at this code, seems like `Mobility::Configuration#new_fallbacks` always return a new instance and whatever is defined with `default_options[:fallbacks] = { ... }` is never applied.

Is this intentional?

I.e. `Mobility.config.new_fallbacks(Mobility.config.default_options[:fallbacks])` returns what I expect from `Mobility.config.new_fallbacks` to return.

**Comment by shioyama** [comment]
That's the expected behaviour. `new_fallbacks` is *passed* the hash of fallbacks (defaulting to `Mobiltiy.default_options[:fallbacks]`) when it is called:

https://github.com/shioyama/mobility/blob/26dcaf011d2076c0d3500763970699c8173ae6f0/lib/mobility/plugins/fallbacks.rb#L148-L152

That said, I think `new_fallbacks` is actually confusing and anyway the method is not really necessary, so I'm thinking of making it private. You should only have to set `fallbacks_generator`, which is a proc that generates the fallbacks instance.

My concern is that for the vast majority of cases, people just want fallbacks to *work* as expected; very few people actually want to define a "fallbacks generator". Ideally Mobility should just do what is expected by default, allowing people to override the default in a sensible way when required.

**Comment by darkcode85** [comment]
If someone want use the `fallbacks: true` option in any model and respect `I18n.fallbacks` values which can be set statically or dynamically  then you can do this:

```
class Blog < ActiveRecord::Base
  extend Mobility
  translates :title, type: :string, fallbacks: true
end
```

then in your `application_controller.rb` you can have a before action similar to this one:
```
  def set_fallbacks
    I18n.fallbacks = set_fallbacks # Method that set the fallbacks hash statically or dynamically

    # Call the fallbacks_generator and pass I18n.fallbacks to respect fallbacks that you have set in `I18n.fallbacks`
    Mobility.config.fallbacks_generator = Proc.new do |fallbacks|
      I18n.fallbacks
    end
  end

This solution works for Mobility v0.4.3

**Comment by shioyama** [comment]
Thanks, but I wouldn't recommend that approach since it's not thread-safe (in the dynamic case).

**Comment by darkcode85** [comment]
@shioyama mmm ok, what you recommend then to get it working for dynamic cases?

**Comment by shioyama** [comment]
@ingemar can you try #180 ? I've made a small change which I think should resolve the issue you raised here.

**Comment by shioyama** [comment]
@darkcode85 Please try the PR in #180 as a quick fix until I merge it:

Gemfile:

```ruby
gem 'mobility', git: 'https://github.com/shioyama/mobility.git', branch: 'default_to_i18n_fallbacks'
```

See if that works for you.

**Comment by shioyama** [comment]
I merged #161 so this is now closed. Let me know if there are any issues.

I'll probably release this either in 0.6, or in 1.0, whichever comes first :smile:

**Comment by shioyama** [comment]
This has been released in version 0.5.0. Please give it a try :smile:

https://rubygems.org/gems/mobility
